// @hubspot/local-dev-lib/errors/apiErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.throwApiUploadError = exports.throwApiError = exports.throwAxiosErrorWithContext = exports.isSpecifiedHubSpotAuthError = exports.isApiUploadValidationError = exports.isGatingError = exports.isMissingScopeError = void 0;
const api_1 = require("../constants/api");
const lang_1 = require("../utils/lang");
const standardErrors_1 = require("./standardErrors");
const i18nKey = 'errors.apiErrors';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isMissingScopeError(err) {
    return (err.isAxiosError &&
        err.status === 403 &&
        !!err.response &&
        err.response.data.category === 'MISSING_SCOPES');
}
exports.isMissingScopeError = isMissingScopeError;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isGatingError(err) {
    return (err.isAxiosError &&
        err.status === 403 &&
        !!err.response &&
        err.response.data.category === 'GATED');
}
exports.isGatingError = isGatingError;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isApiUploadValidationError(err) {
    return (err.isAxiosError &&
        err.status === 400 &&
        !!err.response &&
        !!(err.response?.data?.message || !!err.response?.data?.errors));
}
exports.isApiUploadValidationError = isApiUploadValidationError;
function isSpecifiedHubSpotAuthError(err, { status, category, subCategory }) {
    const statusCodeErr = !status || err.status === status;
    const categoryErr = !category || err.category === category;
    const subCategoryErr = !subCategory || err.subCategory === subCategory;
    return Boolean(err.name === 'HubSpotAuthError' &&
        statusCodeErr &&
        categoryErr &&
        subCategoryErr);
}
exports.isSpecifiedHubSpotAuthError = isSpecifiedHubSpotAuthError;
function parseValidationErrors(responseData = { errors: [], message: '' }) {
    const errorMessages = [];
    const { errors, message } = responseData;
    if (message) {
        errorMessages.push(message);
    }
    if (errors) {
        const specificErrors = errors.map(error => {
            let errorMessage = error.message;
            if (error.errorTokens && error.errorTokens.line) {
                errorMessage = `line ${error.errorTokens.line}: ${errorMessage}`;
            }
            return errorMessage;
        });
        errorMessages.push(...specificErrors);
    }
    return errorMessages;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function logValidationErrors(error) {
    const validationErrorMessages = parseValidationErrors(error?.response?.data);
    if (validationErrorMessages.length) {
        (0, standardErrors_1.throwError)(new Error(validationErrorMessages.join(' '), { cause: error }));
    }
}
/**
 * @throws
 */
function throwAxiosErrorWithContext(
// eslint-disable-next-line @typescript-eslint/no-explicit-any
error, context = {}) {
    const { status } = error;
    const method = error.config?.method;
    const { projectName } = context;
    const isPutOrPost = method === 'put' || method === 'post';
    const action = method && (api_1.HTTP_METHOD_VERBS[method] || api_1.HTTP_METHOD_VERBS.get);
    const preposition = (method && api_1.HTTP_METHOD_PREPOSITIONS[method]) ||
        api_1.HTTP_METHOD_PREPOSITIONS.get;
    const request = context.request
        ? `${action} ${preposition} "${context.request}"`
        : action;
    const messageDetail = request && context.accountId
        ? (0, lang_1.i18n)(`${i18nKey}.messageDetail`, {
            request,
            accountId: context.accountId,
        })
        : 'request';
    const errorMessage = [];
    if (isPutOrPost && context.payload) {
        errorMessage.push((0, lang_1.i18n)(`${i18nKey}.unableToUpload`, { payload: context.payload }));
    }
    const isProjectMissingScopeError = isMissingScopeError(error) && projectName;
    const isProjectGatingError = isGatingError(error) && projectName;
    switch (status) {
        case 400:
            errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.400`, { messageDetail }));
            break;
        case 401:
            errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.401`, { messageDetail }));
            break;
        case 403:
            if (isProjectMissingScopeError) {
                errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.403MissingScope`, {
                    accountId: context.accountId || '',
                }));
            }
            else if (isProjectGatingError) {
                errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.403Gating`, {
                    accountId: context.accountId || '',
                }));
            }
            else {
                errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.403`, { messageDetail }));
            }
            break;
        case 404:
            if (context.request) {
                errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.404Request`, {
                    action: action || 'request',
                    request: context.request,
                    account: context.accountId || '',
                }));
            }
            else {
                errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.404`, { messageDetail }));
            }
            break;
        case 429:
            errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.429`, { messageDetail }));
            break;
        case 503:
            errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.503`, { messageDetail }));
            break;
        default:
            if (status && status >= 500 && status < 600) {
                errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.500Generic`, { messageDetail }));
            }
            else if (status && status >= 400 && status < 500) {
                errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.400Generic`, { messageDetail }));
            }
            else {
                errorMessage.push((0, lang_1.i18n)(`${i18nKey}.codes.generic`, { messageDetail }));
            }
            break;
    }
    if (error?.response?.data?.message &&
        !isProjectMissingScopeError &&
        !isProjectGatingError) {
        errorMessage.push(error.response.data.message);
    }
    if (error?.response?.data?.errors) {
        error.response.data.errors.forEach((err) => {
            errorMessage.push('\n- ' + err.message);
        });
    }
    (0, standardErrors_1.throwError)(new Error(errorMessage.join(' '), { cause: error }));
}
exports.throwAxiosErrorWithContext = throwAxiosErrorWithContext;
/**
 * @throws
 */
function throwApiError(error, context = {}) {
    if (error.isAxiosError) {
        throwAxiosErrorWithContext(error, context);
    }
    (0, standardErrors_1.throwError)(error);
}
exports.throwApiError = throwApiError;
/**
 * @throws
 */
function throwApiUploadError(error, context = {}) {
    if (isApiUploadValidationError(error)) {
        logValidationErrors(error);
    }
    throwApiError(error, context);
}
exports.throwApiUploadError = throwApiUploadError;
